### ISSUE TITLE
ExecuteQueryRetry sets DisableReturnValueCache to true

### ISSUE THREAD

**Issue by jensotto** [SharePoint/PnP-Sites-Core#2175]
#### Category
[x] Bug
[ ] Enhancement

#### Environment
[x] Office 365 / SharePoint Online
[x] SharePoint 2019
[x] SharePoint 2016
[ ] SharePoint 2013

If SharePoint on-premises, what's exact CU version: 

#### Expected or Desired Behavior
When using ClientContextExtensions.ExecuteQueryRetry or ClientContextExtensions.ExecuteQueryRetryAsync (SPOnline) the ClientRuntimeContext passed into the method should not modify the property DisableReturnValueCache.

This should not be changed by the extension method. This is a unwanted side effect that confuses the user. Setting ClientRuntimeContext.DisableReturnValueCache should only be done where it's needed when it's needed. 
Setting ClientRuntimeContext.DisableReturnValueCache to true in ClientContextExtensions.ExecuteQueryRetry has no impact on the query. It needs to be set to true prior to calling a method (list of all methods included below) that caches the return value to have any effect. The result of the current implementation is that all code after the initial call to ClientContextExtensions.ExecuteQueryRetry will have it's return value cache disabled.

I propose to identify all code that really needs this behaviour and use the code contributed in #2174 to temporarily disable the return value cache (PnPCoreUtilities.RunWithDisableReturnValueCache). 

#### Observed Behavior
Whenever a call to ClientContextExtensions.ExecuteQueryRetry or ClientContextExtensions.ExecuteQueryRetryAsync (SPOnline) is made the ClientRuntimeContext.DisableReturnValueCache property will be modified and set to true.

#### Steps to Reproduce

Verify that ClientRuntimeContext.DisableReturnValueCache is false (default).
Call ClientContextExtensions.ExecuteQueryRetry()
ClientRuntimeContext.DisableReturnValueCache has been set to true.

Here is a list of all the methods that uses DisableReturnValueCache:
Microsoft.ProjectServer.Client.CalendarCollection.GetByGuid(Guid) : Calendar
Microsoft.ProjectServer.Client.CalendarCollection.GetById(String) : Calendar
Microsoft.ProjectServer.Client.CalendarExceptionCollection.GetById(Int32) : CalendarException
Microsoft.ProjectServer.Client.CustomFieldCollection.GetByAppAlternateId(String) : CustomField
Microsoft.ProjectServer.Client.CustomFieldCollection.GetByGuid(Guid) : CustomField
Microsoft.ProjectServer.Client.CustomFieldCollection.GetById(String) : CustomField
Microsoft.ProjectServer.Client.DraftAssignmentCollection.GetByGuid(Guid) : DraftAssignment
Microsoft.ProjectServer.Client.DraftAssignmentCollection.GetById(String) : DraftAssignment
Microsoft.ProjectServer.Client.DraftProjectResourceCollection.GetByGuid(Guid) : DraftProjectResource
Microsoft.ProjectServer.Client.DraftProjectResourceCollection.GetById(String) : DraftProjectResource
Microsoft.ProjectServer.Client.DraftTaskCollection.GetByGuid(Guid) : DraftTask
Microsoft.ProjectServer.Client.DraftTaskCollection.GetById(String) : DraftTask
Microsoft.ProjectServer.Client.DraftTaskLinkCollection.GetByGuid(Guid) : DraftTaskLink
Microsoft.ProjectServer.Client.DraftTaskLinkCollection.GetById(String) : DraftTaskLink
Microsoft.ProjectServer.Client.EnterpriseProjectTypeCollection.GetByGuid(Guid) : EnterpriseProjectType
Microsoft.ProjectServer.Client.EnterpriseProjectTypeCollection.GetById(String) : EnterpriseProjectType
Microsoft.ProjectServer.Client.EnterpriseResourceCollection.GetByGuid(Guid) : EnterpriseResource
Microsoft.ProjectServer.Client.EnterpriseResourceCollection.GetById(String) : EnterpriseResource
Microsoft.ProjectServer.Client.EventCollection.GetById(String) : Event
Microsoft.ProjectServer.Client.EventCollection.GetByInt(Int32) : Event
Microsoft.ProjectServer.Client.EventHandlerCollection.GetByGuid(Guid) : EventHandler
Microsoft.ProjectServer.Client.EventHandlerCollection.GetById(String) : EventHandler
Microsoft.ProjectServer.Client.LookupEntryCollection.GetByAppAlternateId(String) : LookupEntry
Microsoft.ProjectServer.Client.LookupEntryCollection.GetByGuid(Guid) : LookupEntry
Microsoft.ProjectServer.Client.LookupEntryCollection.GetById(String) : LookupEntry
Microsoft.ProjectServer.Client.LookupTableCollection.GetByAppAlternateId(String) : LookupTable
Microsoft.ProjectServer.Client.LookupTableCollection.GetByGuid(Guid) : LookupTable
Microsoft.ProjectServer.Client.LookupTableCollection.GetById(String) : LookupTable
Microsoft.ProjectServer.Client.PhaseCollection.GetByGuid(Guid) : Phase
Microsoft.ProjectServer.Client.PhaseCollection.GetById(String) : Phase
Microsoft.ProjectServer.Client.PlanAssignmentCollection.GetByGuid(Guid) : PlanAssignment
Microsoft.ProjectServer.Client.PlanAssignmentCollection.GetById(String) : PlanAssignment
Microsoft.ProjectServer.Client.PlanAssignmentIntervalCollection.GetById(String) : PlanAssignmentInterval
Microsoft.ProjectServer.Client.ProjectCollection.GetByGuid(Guid) : PublishedProject
Microsoft.ProjectServer.Client.ProjectCollection.GetById(String) : PublishedProject
Microsoft.ProjectServer.Client.ProjectDetailPageCollection.GetByGuid(Guid) : ProjectDetailPage
Microsoft.ProjectServer.Client.ProjectDetailPageCollection.GetById(String) : ProjectDetailPage
Microsoft.ProjectServer.Client.ProjectEngagementCollection.GetByGuid(Guid) : ProjectEngagement
Microsoft.ProjectServer.Client.ProjectEngagementCollection.GetById(String) : ProjectEngagement
Microsoft.ProjectServer.Client.PublishedAssignmentCollection.GetByGuid(Guid) : PublishedAssignment
Microsoft.ProjectServer.Client.PublishedAssignmentCollection.GetById(String) : PublishedAssignment
Microsoft.ProjectServer.Client.PublishedProjectResourceCollection.GetByGuid(Guid) : PublishedProjectResource
Microsoft.ProjectServer.Client.PublishedProjectResourceCollection.GetById(String) : PublishedProjectResource
Microsoft.ProjectServer.Client.PublishedTaskCollection.GetByGuid(Guid) : PublishedTask
Microsoft.ProjectServer.Client.PublishedTaskCollection.GetById(String) : PublishedTask
Microsoft.ProjectServer.Client.PublishedTaskLinkCollection.GetByGuid(Guid) : PublishedTaskLink
Microsoft.ProjectServer.Client.PublishedTaskLinkCollection.GetById(String) : PublishedTaskLink
Microsoft.ProjectServer.Client.QueueJobCollection.GetByGuid(Guid) : QueueJob
Microsoft.ProjectServer.Client.QueueJobCollection.GetById(String) : QueueJob
Microsoft.ProjectServer.Client.ResourceEngagementCollection.GetByGuid(Guid) : ResourceEngagement
Microsoft.ProjectServer.Client.ResourceEngagementCollection.GetById(String) : ResourceEngagement
Microsoft.ProjectServer.Client.StageCollection.GetByGuid(Guid) : Stage
Microsoft.ProjectServer.Client.StageCollection.GetById(String) : Stage
Microsoft.ProjectServer.Client.StageCustomFieldCollection.GetByGuid(Guid) : StageCustomField
Microsoft.ProjectServer.Client.StageCustomFieldCollection.GetById(String) : StageCustomField
Microsoft.ProjectServer.Client.StageDetailPageCollection.GetByGuid(Guid) : StageDetailPage
Microsoft.ProjectServer.Client.StageDetailPageCollection.GetById(String) : StageDetailPage
Microsoft.ProjectServer.Client.StatusAssignmentCollection.GetByGuid(Guid) : StatusAssignment
Microsoft.ProjectServer.Client.StatusAssignmentCollection.GetById(String) : StatusAssignment
Microsoft.ProjectServer.Client.StatusAssignmentHistoryLineCollection.GetById(String) : StatusAssignmentHistoryLine
Microsoft.ProjectServer.Client.TimeSheetLineCollection.GetByGuid(Guid) : TimeSheetLine
Microsoft.ProjectServer.Client.TimeSheetLineCollection.GetById(String) : TimeSheetLine
Microsoft.ProjectServer.Client.TimeSheetPeriodCollection.GetByGuid(Guid) : TimeSheetPeriod
Microsoft.ProjectServer.Client.TimeSheetPeriodCollection.GetById(String) : TimeSheetPeriod
Microsoft.ProjectServer.Client.TimeSheetWorkCollection.GetById(String) : TimeSheetWork
Microsoft.ProjectServer.Client.WorkflowDesignerFieldCollection.GetById(String) : WorkflowDesignerField
Microsoft.SharePoint.Client.AlertCollection.GetById(Guid) : Alert
Microsoft.SharePoint.Client.ContentTypeCollection.GetById(String) : ContentType
Microsoft.SharePoint.Client.EventReceiverDefinitionCollection.GetById(Guid) : EventReceiverDefinition
Microsoft.SharePoint.Client.FeatureCollection.GetById(Guid) : Feature
Microsoft.SharePoint.Client.FieldCollection.GetById(Guid) : Field
Microsoft.SharePoint.Client.FieldCollection.GetByInternalNameOrTitle(String) : Field
Microsoft.SharePoint.Client.FieldCollection.GetByTitle(String) : Field
Microsoft.SharePoint.Client.FieldLinkCollection.GetById(Guid) : FieldLink
Microsoft.SharePoint.Client.File.GetLimitedWebPartManager(PersonalizationScope) : LimitedWebPartManager
Microsoft.SharePoint.Client.FileCollection.GetByUrl(String) : File
Microsoft.SharePoint.Client.FileVersionCollection.GetById(Int32) : FileVersion
Microsoft.SharePoint.Client.FolderCollection.GetByUrl(String) : Folder
Microsoft.SharePoint.Client.FormCollection.GetById(Guid) : Form
Microsoft.SharePoint.Client.FormCollection.GetByPageType(PageType) : Form
Microsoft.SharePoint.Client.GroupCollection.GetById(Int32) : Group
Microsoft.SharePoint.Client.GroupCollection.GetByName(String) : Group
Microsoft.SharePoint.Client.LanguageCollection.GetById(Int32) : ClientResult<Language>
Microsoft.SharePoint.Client.List.GetItemById(Int32) : ListItem
Microsoft.SharePoint.Client.List.GetItemByStringId(String) : ListItem
Microsoft.SharePoint.Client.List.GetItemByUniqueId(Guid) : ListItem
Microsoft.SharePoint.Client.List.GetView(Guid) : View
Microsoft.SharePoint.Client.ListCollection.GetById(Guid) : List
Microsoft.SharePoint.Client.ListCollection.GetByTitle(String) : List
Microsoft.SharePoint.Client.ListItemCollection.GetById(Int32) : ListItem
Microsoft.SharePoint.Client.ListItemCollection.GetByStringId(String) : ListItem
Microsoft.SharePoint.Client.ListItemVersionCollection.GetById(Int32) : ListItemVersion
Microsoft.SharePoint.Client.ListTemplateCollection.GetByName(String) : ListTemplate
Microsoft.SharePoint.Client.Navigation.GetNodeById(Int32) : NavigationNode
Microsoft.SharePoint.Client.Publishing.SpotlightChannelCollection.GetById(Int32) : SpotlightChannel
Microsoft.SharePoint.Client.Publishing.SpotlightVideoCollection.GetById(Int32) : SpotlightVideo
Microsoft.SharePoint.Client.Publishing.SubtitleCollection.GetSubtitleFile(String) : ClientResult<Stream>
Microsoft.SharePoint.Client.Publishing.VideoChannelCollection.GetById(Guid) : VideoChannel
Microsoft.SharePoint.Client.Publishing.VideoCollection.GetById(Guid) : VideoItem
Microsoft.SharePoint.Client.Publishing.VideoItem.GetPlaybackUrl(VideoStreamingFormat) : ClientResult<String>
Microsoft.SharePoint.Client.Publishing.VideoThumbnailCollection.GetByIndex(VideoThumbnailChoices) : VideoThumbnail
Microsoft.SharePoint.Client.PushNotificationSubscriberCollection.GetByStoreId(String) : PushNotificationSubscriber
Microsoft.SharePoint.Client.RecycleBinItemCollection.GetById(Guid) : RecycleBinItem
Microsoft.SharePoint.Client.RemoteWeb.GetFolderByServerRelativeUrl(String) : Folder
Microsoft.SharePoint.Client.RoleAssignmentCollection.GetByPrincipalId(Int32) : RoleAssignment
Microsoft.SharePoint.Client.RoleDefinitionCollection.GetById(Int32) : RoleDefinition
Microsoft.SharePoint.Client.RoleDefinitionCollection.GetByName(String) : RoleDefinition
Microsoft.SharePoint.Client.RoleDefinitionCollection.GetByType(RoleType) : RoleDefinition
Microsoft.SharePoint.Client.Site.GetCatalog(Int32) : List
Microsoft.SharePoint.Client.Site.OpenWeb(String) : Web
Microsoft.SharePoint.Client.Taxonomy.LabelCollection.GetByValue(String) : Label
Microsoft.SharePoint.Client.Taxonomy.TaxonomySession.GetTerm(Guid) : Term
Microsoft.SharePoint.Client.Taxonomy.TermCollection.GetById(Guid) : Term
Microsoft.SharePoint.Client.Taxonomy.TermCollection.GetByName(String) : Term
Microsoft.SharePoint.Client.Taxonomy.TermGroupCollection.GetById(Guid) : TermGroup
Microsoft.SharePoint.Client.Taxonomy.TermGroupCollection.GetByName(String) : TermGroup
Microsoft.SharePoint.Client.Taxonomy.TermSet.GetTerm(Guid) : Term
Microsoft.SharePoint.Client.Taxonomy.TermSetCollection.GetById(Guid) : TermSet
Microsoft.SharePoint.Client.Taxonomy.TermSetCollection.GetByName(String) : TermSet
Microsoft.SharePoint.Client.Taxonomy.TermStore.GetGroup(Guid) : TermGroup
Microsoft.SharePoint.Client.Taxonomy.TermStore.GetTerm(Guid) : Term
Microsoft.SharePoint.Client.Taxonomy.TermStore.GetTermSet(Guid) : TermSet
Microsoft.SharePoint.Client.Taxonomy.TermStoreCollection.GetById(Guid) : TermStore
Microsoft.SharePoint.Client.Taxonomy.TermStoreCollection.GetByName(String) : TermStore
Microsoft.SharePoint.Client.TimeZoneCollection.GetById(Int32) : TimeZone
Microsoft.SharePoint.Client.UserCollection.GetByEmail(String) : User
Microsoft.SharePoint.Client.UserCollection.GetByLoginName(String) : User
Microsoft.SharePoint.Client.UserCustomActionCollection.GetById(Guid) : UserCustomAction
Microsoft.SharePoint.Client.UserResource.GetValueForUICulture(String) : ClientResult<String>
Microsoft.SharePoint.Client.ViewCollection.GetById(Guid) : View
Microsoft.SharePoint.Client.ViewCollection.GetByTitle(String) : View
Microsoft.SharePoint.Client.Web.GetCatalog(Int32) : List
Microsoft.SharePoint.Client.Web.GetFileByServerRelativeUrl(String) : File
Microsoft.SharePoint.Client.Web.GetFolderByServerRelativeUrl(String) : Folder
Microsoft.SharePoint.Client.Web.GetList(String) : List
Microsoft.SharePoint.Client.Web.GetUserById(Int32) : User
Microsoft.SharePoint.Client.WebParts.WebPartDefinitionCollection.GetByControlId(String) : WebPartDefinition
Microsoft.SharePoint.Client.WebParts.WebPartDefinitionCollection.GetById(Guid) : WebPartDefinition
Microsoft.SharePoint.Client.WebTemplateCollection.GetByName(String) : WebTemplate
Microsoft.SharePoint.Client.Workflow.WorkflowAssociationCollection.GetByName(String) : WorkflowAssociation
Microsoft.SharePoint.Client.Workflow.WorkflowTemplateCollection.GetById(Guid) : WorkflowTemplate
Microsoft.SharePoint.Client.Workflow.WorkflowTemplateCollection.GetByName(String) : WorkflowTemplate

**Comment by jansenbe** [comment]
Hi @jensotto ,

The reason why we set DisableReturnValueCache to true is because of (internal) recommendations. Below is a literal comment from the folks that developed this:

```CSharp
        /// <summary>
        /// Whether to disable caching for method return values.
        /// It is recommended to always set DisableReturnValueCache to true.
        /// The default value is false for backwards compatibility.
        /// </summary>
        /// <remarks>
        /// By default, the client runtime caches the return value for certain API methods
        /// so that additional requests for the same server object will return the same
        /// ClientObject instance. This optimization avoids confusion about object identities
        /// and reduces protocol overhead, however it relies on an assumption that the
        /// server object will not be deleted or changed during the lifetime of the ClientContext.
        /// If that assumption is invalidated, then the API method may return an incorrect result
        /// unless DisableReturnValueCache is set to true.
        /// </remarks>
```

I understand that it impacts the group provisioning scenario in a certain way and let's find a solution for that problem without adjusting the DisableReturnValueCache setting.

**Comment by jensotto** [comment]
This should have been better documented by Microsoft and made available for the public. I did not know the reasoning behind this property as there is no documentation for it. What you quote makes perfect sense, and is more or less what I concluded with.
Since it is recommended to always be set to true I still believe that setting it in the ClientContextExtensions.ExecuteQueryRetry method is not the best location. My claim that this will only affect subsequent calls to the server still holds true. It would be better to do this on every entry points where the context is created or initially used. This will of course require a lot more work than what has currently been implemented.

The only impact it made on the group provisioning is that the cache needs to be disabled, and that is not possible in SharePoint 2013. Still it can't be assumed that it's always disabled by the caller of utility methods, so this property must still be set to false temporarily in those situations.
When internal code is run as part of the provisioning engine it should be OK to assume that the cache is disabled, as long as it actually is disabled. Now it just happens to be disabled because of a side effect of ClientContextExtensions.ExecuteQueryRetry(). Wouldn't it be better to specifically and clearly disable this cache when starting to apply/extract the template, possibly in a cloned context so that it does not affect the existing context.
